compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -198,6 +198,107 @@ pub enum TyKind<'tcx> {
     Error(DelaySpanBugEmitted),
 }
 
+// try to use less hash operations to speed up hashing compared to a derived
+// implementation
+impl<'tcx> std::hash::Hash for TyKind<'tcx> {
+    fn hash<H: std::hash::Hasher>(&self, h: &mut H) {
+        let disc = std::intrinsics::discriminant_value(self) as usize;
+        match self {
+            Self::Bool | Self::Char | Self::Str | Self::Never | Self::Error(_) => disc.hash(h),
+            Self::Int(i) => {
+                (disc | ((std::intrinsics::discriminant_value(i) as usize) << 5)).hash(h);
```

Comment:
Maybe the 5 should be extracted to a `const` with a comment that it needs to be updated if the number of variants changes?

Reply:
Agreed. I wanted a perf run before I take care of such details though.

Reply:
Even better, you could just use [`std::mem::variant_count`](https://doc.rust-lang.org/nightly/std/mem/fn.variant_count.html).